ergo-lib/src/chain/block_header.rs

```diff
@@ -1,102 +1,136 @@
 //! Block header
+use num_bigint::BigInt;
 
-use std::convert::TryFrom;
-use std::convert::TryInto;
-
-use ergotree_ir::mir::header::PreHeader;
+use ergotree_ir::chain::{
+    block_id::BlockId,
+    digest::{ADDigest, Digest32},
+    header::Header,
+    votes::Votes,
+};
 use ergotree_ir::sigma_protocol::dlog_group;
 #[cfg(feature = "json")]
 use serde::{Deserialize, Serialize};
 
-use super::Base16DecodedBytes;
-use super::Base16EncodedBytes;
-use super::Digest32;
-use thiserror::Error;
-
-/// Block id
+/// New-type wrapper for deserializing the remote `Header` type.
 #[cfg_attr(feature = "json", derive(Serialize, Deserialize))]
-#[derive(PartialEq, Eq, Debug, Clone)]
-pub struct BlockId(Digest32);
+#[derive(Debug, Clone)]
+pub struct HeaderJsonHelper(
+    #[cfg_attr(feature = "json", serde(with = "BlockHeaderRef"))] pub Header,
+);
 
-/// Votes for changing system parameters
+/// Block header reference to `Header` type in ergotree-ir.
+///
+/// We do not implement serde traits in ergotree-ir, but they are requested
+/// by crate users. So we split definition into the main one in ergotree-ir and
+/// the one which implements `Serialize` and `Deserialize` (which is here).
 #[cfg_attr(feature = "json", derive(Serialize, Deserialize))]
-#[cfg_attr(
-    feature = "json",
-    serde(
-        into = "Base16EncodedBytes",
-        try_from = "crate::chain::json::block_header::VotesEncodingVariants"
-    )
-)]
+#[cfg_attr(feature = "json", serde(remote = "Header"))]
 #[derive(PartialEq, Eq, Debug, Clone)]
-pub struct Votes(pub [u8; 3]);
-
-/// Votes errors
-#[derive(Error, Debug)]
-pub enum VotesError {
-    /// Invalid byte array size
-    #[error("Votes: Invalid byte array size ({0})")]
-    InvalidSize(#[from] std::array::TryFromSliceError),
+pub(in crate::chain) struct BlockHeaderRef {
```

Comment:
To keep naming consistent `HeaderSerde` should be a better fit. We already have a few "remote" types for serde with `Serde` suffix.